docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,247 @@
+creation-date: 2021-03-10
+last-updated: 2021-03-10
+status: implementable
+---
+
+# Opt-in Autoscaling from Zero
+
+## Table of Contents
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Glossary](#glossary)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals/Future Work](#non-goalsfuture-work)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [Story 1](#story-1)
+      - [Story 2](#story-2)
+    - [Security Model](#security-model)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan](#test-plan-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+* **Node Group** This term has special meaning within the cluster autoscaler, it refers to collections
+  of nodes, and related physical hardware, that are organized within the autoscaler for scaling operations.
+  These node groups do not have a direct relation to specific CRDs within Kubernetes, and may be handled
+  differently by each autoscaler cloud implementation. In the case of Cluster API, node groups correspond
+  directly to MachineSets and MachineDeployments that are marked for autoscaling.
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+The [Kubernetes cluster autoscaler](https://github.com/kubernetes/autoscaler) currently supports
+scaling on Cluster API deployed clusters. One feature that is missing from this integration is
+the ability to scale down to, and up from, a MachineSet or MachineDeployment with zero replicas.
+
+This proposal defines an opt-in mechanism whereby Cluster API users and cloud providers can define
+the specific resource requirements for each Infrastructure Machine Temalate they create. These
+requirements are then utilized by the cluster autoscaler to make decisions about the number
+of nodes to add during scale up operations.
+
+## Motivation
+
+Allowing the cluster autoscaler to scale down its node groups to zero replicas is a common feature
+implemented for many of the integrated cloud providers. It is a popular feature that has been
+requested for Cluster API on multiple occasions. This feature empowers users to reduce their
+operational resource needs, and likewise reduce their operating costs.
+
+Given that Cluster API is an abstraction point that provides access to multiple concrete cloud
+implementations, this feature might not make sense in all scenarios. To accomodate the wide
+range of deployment options in Cluster API, the scale to zero feature will be optional for
+users and cloud providers.
+
+### Goals
+
+- Provide capability for Cluster API MachineSets and MachineDeployments to scale from and to zero replicas.
+- Provide a mechanism for users to override the defaults for any given MachineSet or MachineDeployment.
```

Comment:
which defaults is this referring to?

Reply:
i guess "defaults" is ambiguous here. i mean to say that there will be a mechanism for users to override the values provided in the infrastructure machine template.